File: platform/app/models/template.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Optional

from sqlalchemy import Boolean, DateTime, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class TaskTemplateModel(Base):
    __tablename__ = "task_templates"

    id: Mapped[str] = mapped_column(
        String(36), primary_key=True, default=lambda: str(uuid.uuid4())
    )
    name: Mapped[str] = mapped_column(String(100), unique=True, nullable=False)
    display_name: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    stages: Mapped[str] = mapped_column(Text, nullable=False, default="[]")
    gates: Mapped[str] = mapped_column(Text, nullable=False, default="[]")
    is_builtin: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime, nullable=False, server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, nullable=False, server_default=func.now(), onupdate=func.now()
    )

File: web/src/types/template.ts
export interface StageDefinition {
  name: string;
  agent_role: string;
  order: number;
}

export interface GateDefinition {
  after_stage: string;
  type: string;
}

export interface TaskTemplate {
  id: string;
  name: string;
  display_name: string;
  description: string | null;
  stages: StageDefinition[];
  gates: GateDefinition[];
  is_builtin: boolean;
  created_at: string;
  updated_at: string;
}

export interface TemplateListResponse {
  items: TaskTemplate[];
  total: number;
}
